sgkit/stats/genedrop.py

```diff
@@ -0,0 +1,345 @@
+    will have identical inheritance patterns if initialized with identical
+    random seeds when using an array of seeds (see the example).
+
+    Examples
+    --------
+    Dataset with founder genotypes
+
+    >>> import sgkit as sg
+    >>> import numpy as np
+    >>> ds = sg.simulate_genotype_call_dataset(n_variant=3, n_sample=5, n_allele=4)
+    >>> ds["parent_id"] = ["samples", "parents"], [
+    ...     [".", "."],
+    ...     [".", "."],
+    ...     ["S0", "S1"],
+    ...     ["S0", "S1"],
+    ...     ["S2", "S3"],
+    ... ]
+    >>> ds.call_genotype.data[:] = -1
+    >>> ds.call_genotype.data[:,0:2] = np.arange(4).reshape(2,2)
+    >>> sg.display_genotypes(ds) # doctest: +NORMALIZE_WHITESPACE
+    samples    S0   S1   S2   S3   S4
+    variants
+    0         0/1  2/3  ./.  ./.  ./.
+    1         0/1  2/3  ./.  ./.  ./.
+    2         0/1  2/3  ./.  ./.  ./.
+
+    Simulation with random seed
+
+    >>> sim = sg.simulate_genedrop(ds, merge=False, seed=1)
+    >>> sim["sample_id"] = ds["sample_id"]
+    >>> sim["variant_position"] = ds["variant_position"]
+    >>> sim["variant_allele"] = ds["variant_allele"]
+    >>> sg.display_genotypes(sim) # doctest: +NORMALIZE_WHITESPACE
+    samples    S0   S1   S2   S3   S4
+    variants
+    0         0/1  2/3  0/3  1/3  0/3
+    1         0/1  2/3  0/3  0/3  0/3
+    2         0/1  2/3  0/2  0/3  2/0
+
+    Simulation with seed per variant (including duplicates)
+
+    >>> seeds = np.array([0,0,1], 'uint32')
+    >>> sim = sg.simulate_genedrop(ds, merge=False, seed=seeds)
+    >>> sim["sample_id"] = ds["sample_id"]
+    >>> sim["variant_position"] = ds["variant_position"]
+    >>> sim["variant_allele"] = ds["variant_allele"]
+    >>> sg.display_genotypes(sim) # doctest: +NORMALIZE_WHITESPACE
+    samples    S0   S1   S2   S3   S4
+    variants
+    0         0/1  2/3  1/2  0/3  2/3
+    1         0/1  2/3  1/2  0/3  2/3
+    2         0/1  2/3  0/2  1/3  2/3
+
+    References
+    ----------
+    [1] Jean W. MacCluer, John L. VanderBerg. Bruce Read and Oliver A. Ryder 1986.
+    "Pedigree analysis by computer simulation." Zoo Biology 5: 147-160.
+
+    [2] - Richard J. Kerr, Li Li, Bruce Tier, Gregory W. Dutkowski and Thomas A. McRae 2012.
+    "Use of the numerator relationship matrix in genetic analysis of autopolyploid species."
+    Theoretical and Applied Genetics 124: 1271-1282.
+
+    [3] - Matthew G. Hamilton and Richard J. Kerr 2017.
+    "Computation of the inverse additive relationship matrix for autopolyploid
+    and multiple-ploidy populations." Theoretical and Applied Genetics 131: 851-860.
+    """
+    ds = define_variable_if_absent(ds, variables.parent, parent, parent_indices)
+    variables.validate(
+        ds, {parent: variables.parent_spec, call_genotype: variables.call_genotype_spec}
+    )
+    gt = da.asarray(ds[call_genotype].data)
+    parent = da.asarray(ds[parent].data, chunks=ds[parent].shape)
+    if hasattr(seed, "__len__"):
```

Comment:
@jeromekelleher, the code relevant to deterministic results starts here. We generate a random seed per variant at call time (unless provided by the user). This is used at evaluation time to determine the inheritance pattern at each variant. The numba random generator is thread safe so this should be sufficient and appears to be based on testing.